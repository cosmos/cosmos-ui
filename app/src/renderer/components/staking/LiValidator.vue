<template lang='pug'>
li.li-validator(:class='styles')
  router-link(:to="{ name: 'validator', params: { validator: validator.id }}")
    .li-validator__value.name
      span.validator-profile__status(v-bind:class="statusColor" v-tooltip.top="status")
      img.avatar(v-if="validator.keybase" :src="validator.keybase.avatarUrl" width="48" height="48")
      img.avatar(v-else src="~assets/images/validator-icon.svg" width="48" height="48")
      .vert
        .top {{ validator.description.moniker }}
        short-bech32(:address="validator.operator_address")
    .li-validator__value.your-votes
      span {{ yourVotes.isLessThan(0.01) && yourVotes.isGreaterThan(0) ? '< ' + num.full(0.01) : num.full(yourVotes) }}
    .li-validator__value.your-rewards
      span n/a
    .li-validator__break: span
    .li-validator__value.percent_of_vote
      span {{ validator.percent_of_vote ? validator.percent_of_vote : `n/a` }}
    .li-validator__value.uptime
      // add .green .yellow or .red class to this span to trigger inidication by color
      span {{ uptime }}
    .li-validator__value.commission
      span {{ commission }}
    .li-validator__value.slashes
      // add .green .yellow or .red class to this span to trigger inidication by color
      span {{ slashes }}
</template>

<script>
import { mapGetters } from "vuex"
import num from "scripts/num"
import { calculateTokens, ratToBigNumber } from "scripts/common"
import ShortBech32 from "common/ShortBech32"
import BigNumber from "bignumber.js"
export default {
  name: `li-validator`,
  props: [`validator`, `disabled`],
  components: {
    ShortBech32
  },
  computed: {
    ...mapGetters([`delegates`, `committedDelegations`]),
    slashes() {
      return `n/a` //TODO: add slashes
    },
    commission() {
      return `${this.num.pretty(this.validator.commission.rate)}%`
    },
    uptime() {
      let rollingWindow = 10000 // param of slashing period
      let info = this.validator.signing_info
      if (info) {
        // uptime in the past 10k blocks
        let uptimeRollingWindow = info.signed_blocks_counter / rollingWindow
        return `${this.num.pretty(uptimeRollingWindow * 100)}%`
      }
      return `n/a`
    },
    // TODO uncomment when distribution is done
    // yourRewards() {
    //   if (
    //     this.committedDelegations[this.validator.id] &&
    //     this.committedDelegations[this.validator.id].isValidator &&
    //     this.validator.proposer_reward_pool > 0
    //   ) {
    //     let myShares = calculateShares(
    //       this.validator,
    //       this.committedDelegations[this.validator.id]
    //     )
    //     let shareRatio = myShares.div(parseValidatorShares(this.validator))
    //     let rewardsInShares = shareRatio.times(
    //       this.validator.proposer_reward_pool
    //     )
    //     let rewardsInTokens = calculateTokens(rewardsInShares)
    //     return this.num.full(rewardsInTokens).toString()
    //   } else return "0"
    // },
    yourVotes() {
      return this.committedDelegations[this.validator.id]
        ? calculateTokens(
            this.validator,
            this.committedDelegations[this.validator.id]
          )
        : BigNumber(0)
    },
    styles() {
      let value = ``
      if (this.validator.isValidator) value += `li-validator-validator `
      return value
    },
    delegateType() {
      return this.validator.revoked
        ? `Revoked`
        : this.validator.isValidator
          ? `Validator`
          : `Candidate`
    },
    powerRatio() {
      return ratToBigNumber(this.validator.tokens)
        .div(this.delegates.globalPower)
        .toNumber()
    },
    // TODO enable once we decide on limits
    // powerRatioLevel() {
    //   if (this.powerRatio < 0.01) return "green"
    //   if (this.powerRatio < 0.03) return "yellow"
    //   else if (this.powerRatio >= 0.03) return "red"
    // },
    status() {
      // status: jailed
      if (this.validator.revoked)
        return `This validator has been jailed and is not currently validating`

      // status: candidate
      if (parseFloat(this.validator.voting_power) === 0)
        return `This validator has declared candidacy but does not have enough voting power yet`

      // status: validator
      return `This validator is actively validating`
    },
    statusColor() {
      // status: jailed
      if (this.validator.revoked) return `red`

      // status: candidate
      if (parseFloat(this.validator.voting_power) === 0) return `yellow`

      // status: validator
      return `green`
    }
  },
  data: () => ({ num })
}
</script>

<style lang="stylus">
@require '~variables'

.li-validator
  margin 0.5rem 0rem 0.5rem 2rem

.li-validator a
  display flex
  align-items center
  padding 1rem
  background-color var(--app-fg)
  border-radius 0.25rem
  border 1px solid var(--bc-dim)

  &:hover
    background var(--hover-bg)

  .li-validator__value:not(:first-of-type) span
    color var(--dim)
    background-color var(--white-fade-1)
    border 1px solid var(--white-fade-2)
    border-radius 4px
    display block
    width 100%
    margin 0 0.5rem
    font-size sm
    line-height sm
    text-align right
    padding 0.25rem 0.5rem

    &.red
      background-color rgba(209, 2, 0, 0.15)
      border solid 0.5px rgba(209, 2, 0, 0.25)
      color #ff0200

    &.yellow
      background-color rgba(255, 149, 2, 0.15)
      border solid 0.5px rgba(255, 149, 2, 0.25)
      color #ff9502

    &.green
      background-color rgba(46, 164, 45, 0.15)
      border solid 0.5px rgba(46, 164, 45, 0.25)
      color #2ea42d

  .li-validator__break
    flex 0
    display flex
    align-items center
    min-width 1

    span
      margin 0 0.5rem
      width 1px
      background-color var(--white-fade-1)
      height 2rem

  .li-validator__value
    flex 1
    display flex
    align-items center
    min-width 0

    &.name
      flex 3

      a
        display flex

        img
          border-radius 100%
          margin-right 1rem

  .vert
    display flex
    flex-direction column
    color var(--bright)
    padding-left 1rem

    .top
      padding-bottom 0.5rem
      line-height 1rem
      font-size 1rem
      font-weight 500

    .bottom
      font-size sm
      line-height sm
      color var(--dim)

  .li-validator__icon
    width 1.5rem
    display flex
    align-items center
    justify-content center

    img, span
      height 1rem
      width 1rem

    &.bar
      position relative

      span
        display block
        position absolute
        top 0
        left 0
        z-index z(listItem)
        line-height 3rem
        color var(--txt)

      .bar
        height 1.5rem
        position relative
        left -0.25rem
        background var(--accent-alpha)

    &.checkbox
      justify-content center
      cursor pointer

    span
      white-space nowrap
      overflow hidden
      text-overflow ellipsis
      padding-right 1rem

ol
  list-style none
  counter-reset counter

ol li
  counter-increment counter
  position relative
  height 82px

ol li:before
  content counter(counter) ''
  color var(--dim)
  font-size sm
  position absolute
  display block
  left -1.5rem
  line-height 82px
  height 100%
</style>
