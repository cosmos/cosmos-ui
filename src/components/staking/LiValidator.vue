<template>
  <tr
    class="li-validator"
    :data-moniker="validator.description.moniker"
    @click="
      $router.push({
        name: 'validator',
        params: { validator: validator.operator_address }
      })
    "
  >
    <td class="data-table__row__info">
      <ApolloQuery
        :query="ValidatorProfile"
        :variables="{ address: validator.operator_address }"
        :update="validatorProfileResultUpdate"
      >
        <template v-slot="{ result: { loading, error, data: keybase } }">
          <Avatar
            v-if="!keybase || !keybase.avatarUrl || loading || error"
            class="li-validator-image"
            alt="generic validator logo - generated avatar from address"
            :address="validator.operator_address"
          />
          <img
            v-else-if="keybase && keybase.avatarUrl"
            :src="keybase.avatarUrl"
            class="li-validator-image"
            :alt="`validator logo for ` + validator.description.moniker"
          />
        </template>
      </ApolloQuery>
      <div class="validator-info">
        <h3 class="li-validator-name">
          {{ validator.description.moniker }}
        </h3>
        <div v-if="validator.my_delegations > 0">
          <h4>
            {{ validator.my_delegations | atoms | shortDecimals }}
          </h4>
          <h5 v-if="validator.rewards > 0">
            +{{ validator.rewards | atoms | shortDecimals }}
          </h5>
        </div>
      </div>
    </td>
    <td :class="{ 'hide-xs': showOnMobile !== 'expectedReturns' }">
      {{
        validator.expectedReturns ? percent(validator.expectedReturns) : `--`
      }}
    </td>
    <td :class="{ 'hide-xs': showOnMobile !== 'voting-power' }">
      {{ validator.tokens ? percentOfVotingPower : `--` }}
    </td>
  </tr>
</template>

<script>
import { mapState } from "vuex"
import { percent, shortDecimals, atoms } from "scripts/num"
import Avatar from "common/Avatar"
import BN from "bignumber.js"
import { ValidatorProfile, validatorProfileResultUpdate } from "src/gql"

export default {
  name: `li-validator`,
  components: {
    Avatar
  },
  filters: {
    atoms,
    shortDecimals
  },
  props: {
    validator: {
      type: Object,
      required: true
    },
    showOnMobile: {
      type: String,
      default: () => "returns"
    }
  },
  data: () => ({
    ValidatorProfile
  }),
  computed: {
    ...mapState([`pool`]),
    percentOfVotingPower() {
      return percent(
        BN(this.validator.tokens)
          .div(this.pool.pool.bonded_tokens)
          .toFixed(4)
      )
    }
  },
  methods: {
    shortDecimals,
    atoms,
    percent,
    validatorProfileResultUpdate
  }
}
</script>
<style scoped>
.li-validator {
  padding: 0.5rem 1rem;
  margin-bottom: 0.25rem;
  border-bottom: 1px solid var(--bc-dim);
  border-radius: 0.25rem;
}

.li-validator:last-child {
  border-bottom: none;
}

.validator-info {
  display: flex;
  flex-direction: column;
  padding-left: 1rem;
  text-overflow: ellipsis;
}

.li-validator h4,
.li-validator h5 {
  font-size: var(--sm);
  display: inline-block;
}

.li-validator h5 {
  padding-left: 0.5rem;
  color: var(--success);
}

.li-validator:hover {
  cursor: pointer;
  background: var(--hover-bg);
  color: var(--bright);
}

.li-validator-name {
  font-size: 1rem;
  line-height: 18px;
  font-weight: 500;
  color: var(--bright);
  display: inline-block;
}

.li-validator-image {
  border-radius: 0.25rem;
  height: 2.5rem;
  width: 2.5rem;
  border: 1px solid var(--bc-dim);
}
</style>
