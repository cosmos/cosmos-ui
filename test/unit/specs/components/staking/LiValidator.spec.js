import { shallowMount, createLocalVue } from "@vue/test-utils"
import LiValidator from "renderer/components/staking/LiValidator"

const localVue = createLocalVue()
localVue.directive(`tooltip`, () => { })

describe(`LiValidator`, () => {
  let wrapper, $store

  const validator = {
    operator_address: `cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au`,
    pub_key: `cosmosvalpub1234`,
    revoked: false,
    tokens: `14`,
    delegator_shares: `14`,
    description: {
      website: `www.monty.ca`,
      details: `Mr Mounty`,
      moniker: `mr_mounty`,
      country: `Canada`
    },
    status: 2,
    bond_height: `0`,
    bond_intra_tx_counter: 6,
    proposer_reward_pool: null,
    commission: {
      rate: `0.05`,
      max_rate: `0.1`,
      max_change_rate: `0.005`,
      update_time: `1970-01-01T00:00:00Z`
    },
    prev_bonded_shares: `0`,
    voting_power: `10`,
    percent_of_vote: `22%`,
    signing_info: {
      start_height: 0,
      index_offset: 465400,
      jailed_until: `1970-01-01T00:00:00Z`,
      missed_blocks_counter: 122
    }
  }

  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters: {
        delegates: { delegates: [], globalPower: 9000 },
        committedDelegations: {},
        distribution: {
          rewards: {
          }
        },
        session: {
          signedIn: true
        },
        bondDenom: `stake`,
        lastHeader: ``
      }
    }

    wrapper = shallowMount(LiValidator, {
      localVue,
      propsData: {
        validator,
        disabled: false
      },
      mocks: {
        $store
      },
      stubs: [`router-link`]
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it(`should calculate the validator's power ratio`, () => {
    const ratio = wrapper.vm.validator.tokens / wrapper.vm.delegates.globalPower
    expect(wrapper.vm.powerRatio).toBe(ratio)
  })

  it(`should show the voting power`, () => {
    expect(wrapper.html()).toContain(`22%`)
  })

  it(`should show the validator status`, () => {
    expect(wrapper.vm.status).toBe(`This validator is actively validating`)
    // Jailed
    wrapper.setProps({
      validator: Object.assign({}, validator, {
        revoked: true
      })
    })
    expect(wrapper.vm.status).toBe(
      `This validator has been jailed and is not currently validating`
    )
    // Is not a validator
    wrapper.setProps({
      validator: Object.assign({}, validator, {
        revoked: false,
        voting_power: 0
      })
    })
    expect(wrapper.vm.status).toBe(
      `This validator does not have enough voting power yet and is inactive`
    )
  })

  it(`should show the validator status with color`, () => {
    expect(wrapper.vm.statusColor).toBe(`green`)
    // Jailed
    wrapper.setProps({
      validator: Object.assign({}, validator, {
        revoked: true
      })
    })
    expect(wrapper.vm.statusColor).toBe(`red`)
    // Is not a validator
    wrapper.setProps({
      validator: Object.assign({}, validator, {
        revoked: false,
        voting_power: 0
      })
    })
    expect(wrapper.vm.statusColor).toBe(`yellow`)
  })

  it(`should show the validator's uptime`, () => {
    expect(wrapper.vm.uptime).toBe(`98.78%`)
    expect(wrapper.html()).toContain(`98.78%`)

    wrapper.setProps({
      validator: Object.assign({}, validator, {
        signing_info: null
      })
    })

    expect(wrapper.vm.uptime).toBe(`--`)
  })

  it(`should show the validator's commission`, () => {
    expect(wrapper.html()).toContain(wrapper.vm.validator.commission.rate)
  })

  it(`should show the type of the candidate`, () => {
    wrapper.setProps({
      validator: Object.assign({}, validator, {
        revoked: false,
        isValidator: false
      })
    })
    expect(wrapper.vm.delegateType).toBe(`Candidate`)
    wrapper.setProps({
      validator: Object.assign({}, validator, {
        revoked: false,
        isValidator: true
      })
    })
    expect(wrapper.vm.delegateType).toBe(`Validator`)
    wrapper.setProps({
      validator: Object.assign({}, validator, {
        revoked: true,
        isValidator: false
      })
    })
    expect(wrapper.vm.delegateType).toBe(`Revoked`)
    wrapper.setProps({
      validator: Object.assign({}, validator, {
        revoked: true,
        isValidator: true
      })
    })
    expect(wrapper.vm.delegateType).toBe(`Revoked`)
  })

  it(`shows rewards`, () => {
    $store.getters.distribution.rewards = {
      [validator.operator_address]: {
        stake: 1230000000
      }
    }

    expect(wrapper.find(`.li-validator__rewards`).html()).toContain(`123.0000â€¦`)
  })

  it(`works if user is not signed in`, () => {
    $store.getters.session.signedIn = false

    expect(wrapper.vm.$el).toMatchSnapshot()
  })
})
