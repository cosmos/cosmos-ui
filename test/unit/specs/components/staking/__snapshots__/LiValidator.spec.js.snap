// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiValidator has the expected html structure 1`] = `
<tr
  class="data-table__row li-validator"
>
  <td
    class="data-table__row__info"
  >
    <img
      class="data-table__row__info__image data-table__row__info__image--no-img"
      height="48"
      src="~assets/images/validator-icon.svg"
      width="48"
    />
     
    <div
      class="data-table__row__info__container"
    >
      <span
        class="data-table__row__info__container__status green"
      />
       
      <a
        class="data-table__row__info__container__name"
        href="#/staking/validators/cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au"
      >
        herr_schmidt_revoked
      </a>
       
      <div
        class="short-bech32 data-table__row__info__container__description"
      >
        <div
          class="address"
          id="address"
        >
          
    cosmosvaladdr…n3au
  
        </div>
         
        <div
          class="copied"
        >
          <i
            class="material-icons"
          >
            check
          </i>
          <span>
            Copied
          </span>
        </div>
      </div>
    </div>
  </td>
   
  <td
    class="li-validator__delegated-steak"
  >
    
    0.0000…
  
  </td>
   
  <td
    class="li-validator__rewards"
  >
    n/a
  </td>
   
  <td
    class="li-validator__voting-power"
  >
    
    22%
  
  </td>
   
  <td
    class="li-validator__uptime"
  >
    98.78%
  </td>
   
  <td
    class="li-validator__commission"
  >
    0.05%
  </td>
   
  <td
    class="li-validator__slashes"
  >
    n/a
  </td>
</tr>
`;
